f2z-relay: §5.3 says a tls-exporter relay MUST refuse a zero binding; export() returns zeros and the connection proceeds

Found reviewing `f2z-relay/src/tls.rs` and `listener.rs` on `c7a1554` — files #628 landed that had not been reviewed since. Baseline `cargo test -p f2z-relay -p f2z-relay-proto`: **EXIT=0, 255 passed**.

## The rule

`WIRE.md` §5.3, lines 574-577, is a relay-side MUST and it is unusually explicit about why:

> A relay in `tls-exporter` mode whose exporter output is 32 zero bytes **MUST treat that as failure to obtain a binding and refuse the connection.** Zero is the `none` sentinel; accepting it in both modes would make them identical in the signed transcript.

## What the code does

`tls.rs`'s `export()` returns the sentinel and lets the connection proceed:

```rust
match connection.export_keying_material(output, EXPORTER_LABEL, None) {
    Ok(bytes) => f2z_codec::types::ChannelBinding::new(bytes),
    Err(_) => {
        crate::log_warn!("TLS exporter unavailable; this connection has no channel binding");
        output = [0u8; EXPORTER_LEN];
        f2z_codec::types::ChannelBinding::new(output)
    }
}
```

`listener.rs:158-166` takes whatever comes back and completes the upgrade — there is no check on the value:

```rust
let binding = { let (_io, connection) = tls.get_ref(); crate::tls::export(connection) };
let socket = tokio_tungstenite::accept_hdr_async(tls, check_upgrade).await.ok()?;
Some((wrap(socket), binding))
```

And the relay really is in `tls-exporter` mode on that path: `caps.rs:138` derives the published mode from `config.tls_enabled()`, which is what put an acceptor there in the first place. So the connection proceeds with the exact `ChannelBinding` value that `listener.rs:176` uses deliberately for the *plaintext* path — the two modes become identical in the signed transcript, which is the sentence §5.3 wrote the MUST to prevent.

The `Ok` branch is unchecked too. §5.3's rule is about the exporter's **output**, not about the error: a successful all-zero return would also be passed through. That case has probability 2^-256 and is not why I am filing, but the guard that covers one covers both.

## The module comment contradicts the spec

```rust
/// Returns 32 zero bytes when there is no TLS session, which is the value §5.3
/// requires in `channel_binding_mode: none` — the same constant both ends use,
/// so the transcript still verifies and the binding simply provides nothing.
```

§5.3 requires zeros in `none` **mode**, and forbids accepting them in `tls-exporter` mode. "The same constant both ends use" is the assumption that does not hold here: the peer is a TLS client whose own exporter succeeded, so it derives a real binding, not zeros.

## Severity — stated plainly, because it is lower than it first looks

**This is fail-closed in practice, and I could not construct a case where an unbound session is silently accepted end to end.** If the relay's exporter fails and the client's succeeds, the client signs against its real binding, the relay rebuilds the transcript with zeros, and every signature fails `ERR_BAD_SIGNATURE`. The connection is useless rather than unprotected.

So the value of fixing it is conformance and a clear failure instead of an obscure one — which is exactly the trade `f2z-relay-proto/src/hello.rs:240-243` already articulates for the client side:

> a signature made under that disagreement fails later, obscurely. Fail now, clearly.

That client-side check (`hello.rs:248-255`) is a **different rule**: it validates the relay's *claimed* mode against the binding the client derived, and it catches the mirror case (client derived nothing, relay claims an exporter). It does not and cannot make the relay refuse its own failed exporter. I checked it before filing precisely because a guard in another layer usually means the finding is "duplicated" rather than "absent"; here it is genuinely a different rule on the other end of the wire.

`export_keying_material` failing on a completed TLS 1.3 handshake is also unlikely. §5.3 anticipated it anyway and wrote a MUST, and a warning log is not a refusal.

## Fix

Make `export` fallible and let the caller drop the connection:

```rust
pub fn export(connection: &ServerConnection) -> Option<ChannelBinding> {
    let mut output = [0u8; EXPORTER_LEN];
    let bytes = connection.export_keying_material(output, EXPORTER_LABEL, None).ok()?;
    // §5.3: zero is the `none` sentinel. In tls-exporter mode it is a failure
    // to obtain a binding, not a binding of zero.
    let binding = ChannelBinding::new(bytes);
    (!binding.is_zero()).then_some(binding)
}
```

and at `listener.rs:163` return `None`, which is already how that function reports "no connection" — the `?` on `accept_hdr_async(...).ok()?` two lines below takes the same path. This matches `acceptor()`'s own stance in the same module, which is the argument for it:

> A relay configured for TLS that cannot serve it must not silently fall back to plaintext: §2.3's insecure path is an explicit act with published consequences, not a degradation.

That is the principle; `export()` is the one place in the file that does the opposite.

A test can drive it without a TLS stack by testing the predicate rather than the handshake: assert that an all-zero exporter output yields no binding, with a non-zero output as the positive control so the refusal is attributable.

## Reviewed in the same pass, no findings

- **TLS 1.2 is structurally unavailable, which is better than a test.** I tried to widen `VERSIONS` to admit it and it does not compile: `error[E0425]: cannot find value TLS12 in module tokio_rustls::rustls::version`. `rs/crates/f2z-relay/Cargo.toml:21` takes `tokio-rustls` with `default-features = false, features = ["ring", "logging"]`, so the `tls12` feature is absent from the build. §2.1's "MUST NOT negotiate TLS 1.2 or below" cannot be violated by editing this file — it would take a Cargo change, in a diff where that is the visible subject.
- `EXPORTER_LABEL` is already pinned to a literal (`tls.rs:161`), so the #657 shape does not apply here.
- The plaintext path's `ChannelBinding::zero()` at `listener.rs:176` is correct and correctly cited.

--- a/rs/crates/f2z-relay/src/listener.rs
+++ b/rs/crates/f2z-relay/src/listener.rs
@@ -158,10 +158,13 @@ async fn handshake(
         Some(acceptor) => {
             let tls = acceptor.accept(stream).await.ok()?;
             // §5.3: the exporter is taken from the completed handshake, before
-            // the WebSocket upgrade consumes the stream.
+            // the WebSocket upgrade consumes the stream. No binding is a
+            // refusal, not a degradation to `none` — this relay publishes
+            // `channel_binding_mode: tls-exporter`, and §5.3 requires the
+            // connection to be refused rather than carry the `none` sentinel.
             let binding = {
                 let (_io, connection) = tls.get_ref();
-                crate::tls::export(connection)
+                crate::tls::export(connection)?
             };
             let socket = tokio_tungstenite::accept_hdr_async(tls, check_upgrade)
                 .await

--- a/rs/crates/f2z-relay/src/tls.rs
+++ b/rs/crates/f2z-relay/src/tls.rs
@@ -130,26 +130,44 @@ fn load_private_key(path: &Path) -> Result<PrivateKeyDer<'static>, TlsError> {
 
 /// §5.3's exporter, taken from a completed handshake.
 ///
-/// Returns 32 zero bytes when there is no TLS session, which is the value §5.3
-/// requires in `channel_binding_mode: none` — the same constant both ends use,
-/// so the transcript still verifies and the binding simply provides nothing.
-#[must_use]
+/// Returns `None` when there is no binding to be had, and the caller drops the
+/// connection. Anything that reaches this function has a TLS session, so
+/// [`crate::caps`] publishes `channel_binding_mode: tls-exporter`, and §5.3 is
+/// explicit about what that costs:
+///
+/// > A relay in `tls-exporter` mode whose exporter output is 32 zero bytes MUST
+/// > treat that as failure to obtain a binding and refuse the connection. Zero
+/// > is the `none` sentinel; accepting it in both modes would make them
+/// > identical in the signed transcript.
+///
+/// So neither a failing exporter nor an exporter that returns the sentinel may
+/// degrade this connection into the `none` mode it does not publish. It is
+/// [`acceptor`]'s stance one layer up, for the same reason: §2.3's insecure
+/// path is an explicit act with published consequences, not a degradation.
 pub fn export(
     connection: &tokio_rustls::rustls::ServerConnection,
-) -> f2z_codec::types::ChannelBinding {
-    let mut output = [0u8; EXPORTER_LEN];
-    match connection.export_keying_material(output, EXPORTER_LABEL, None) {
-        Ok(bytes) => f2z_codec::types::ChannelBinding::new(bytes),
-        // An exporter that cannot be derived from a completed TLS 1.3 handshake
-        // is a `rustls` state this code has no way to repair. Zeros are the
-        // value §5.3 defines for "no binding", so the connection degrades to the
-        // documented weaker mode rather than to a value only one end knows.
-        Err(_) => {
-            crate::log_warn!("TLS exporter unavailable; this connection has no channel binding");
-            output = [0u8; EXPORTER_LEN];
-            f2z_codec::types::ChannelBinding::new(output)
-        }
+) -> Option<f2z_codec::types::ChannelBinding> {
+    let output = [0u8; EXPORTER_LEN];
+    // An exporter that cannot be derived from a completed TLS 1.3 handshake is
+    // a `rustls` state this code has no way to repair.
+    let Ok(bytes) = connection.export_keying_material(output, EXPORTER_LABEL, None) else {
+        crate::log_warn!("TLS exporter unavailable; refusing the connection (WIRE.md §5.3)");
+        return None;
+    };
+    binding_from_exporter(bytes)
+}
+
+/// §5.3's rule about the exporter's **output**, kept separate from the
+/// exporter's *error* so it can be exercised without terminating TLS.
+fn binding_from_exporter(bytes: [u8; EXPORTER_LEN]) -> Option<f2z_codec::types::ChannelBinding> {
+    let binding = f2z_codec::types::ChannelBinding::new(bytes);
+    if binding.is_zero() {
+        crate::log_warn!(
+            "TLS exporter returned the `none` sentinel; refusing the connection (WIRE.md §5.3)"
+        );
+        return None;
     }
+    Some(binding)
 }
 
 #[cfg(test)]
@@ -165,6 +183,25 @@ mod tests {
     #[test]
     fn only_tls_1_3_is_offered() {
         assert_eq!(VERSIONS.len(), 1);
+    }
+
+    #[test]
+    fn a_zero_exporter_output_yields_no_binding() {
+        // §5.3: "A relay in `tls-exporter` mode whose exporter output is 32
+        // zero bytes MUST treat that as failure to obtain a binding and refuse
+        // the connection." `export` returns `None` and `listener::handshake`
+        // drops the connection on it.
+        assert!(binding_from_exporter([0u8; EXPORTER_LEN]).is_none());
+    }
+
+    #[test]
+    fn a_real_exporter_output_yields_that_binding() {
+        // The positive control, so the refusal above is attributable to the
+        // sentinel and not to the guard refusing everything.
+        let mut bytes = [0u8; EXPORTER_LEN];
+        bytes[EXPORTER_LEN - 1] = 1;
+        let binding = binding_from_exporter(bytes).expect("a non-zero exporter output binds");
+        assert_eq!(binding.as_bytes(), &bytes);
     }
 
     #[test]

